### PR TITLE
Restrict template part inserter support to the site editor using filters

### LINKED ISSUE #30668: Template Part block should not be available in the Post Editor inserter
## Description
As of 10.4-RC1, Template Part blocks can be added to the Post Editor in the same way reusable blocks can be added. However, template parts can only be created from the Site Editor.

This issue is related to #30669, which presents the opposite logic for reusable blocks. Original discussion in #30308

## Step-by-step reproduction instructions

1. Go to 'Site Editor'
2. Select any block and click on "Make template part"
3. Go to the Post Editor.
4. Open the inserter and look for the recently created template part.

## Expected behaviour
Template parts should only be available to insert in template editors.

## Actual behaviour
Template parts can be inserted in posts or pages.

## WordPress information
- WordPress version: 5.7
- Gutenberg version: 10.4-RC1
- Are all plugins except Gutenberg deactivated? Yes
- Theme used? TT1 Blocks

### PR DESCRIPTION
**Pull request by adamziel**
Description TBD

Solves https://github.com/WordPress/gutenberg/issues/30668

### CODE REVIEW

File: packages/edit-site/src/index.js [hunk 1]
```diff
@@ -57,7 +58,15 @@ export function initializeEditor( id, settings ) {
 	const reboot = reinitializeEditor.bind( null, target, settings );
 
 	dispatch( blocksStore ).__experimentalReapplyBlockTypeFilters();
+
+	addFilter(
+		'blocks.registerBlockType',
+		'core/edit-site/block-register',
+		addInsertersupportForTemplatePartBlock
+	);
 	registerCoreBlocks();
```
Comment (adamziel) [inline packages/edit-site/src/index.js:67]: If this approach sticks, I can see moving towards a declarative approach. E.g. imagine having an `editor.json` like:
```json
{
    "blockOverrides": {
        "core/template-part": {
            "supports": {
                "inserter": true
            }
        }
    }
}
```

Or a call like

```js
registerCoreBlocksWithOverrides([
    setBlockSupport("core/template-part", "inserter", true)
])
```

Comment (priethor) [reply packages/edit-site/src/index.js:67]: I like this idea of different per editor context. I don't want to derail the conversation as this is worth a separate issue, but how would a 3rd-party block achieve the same functionality (working only the site editor) if they can only pack their `block.json` settings but don't have access to the `editor.json` one?

**Review by priethor — COMMENTED** [review]
Thanks @adamziel !

It behaves as expected, showing the Template Part block only in the Site Editor.

The implementation looks good to me but I'll defer to others to give the final ✅

**Comment by ntsekouras** [comment]
Thanks for the PR Adam!

This approach doesn't solve the problem now, as while we are in `post editor` we can `edit the template` and there (outside the PostContent block) the `Template Part` block should be available.

In general the new API that we need, will have to define what `context` is - and just with this block (Template Part) we can see that the context isn't narrowed to `editors`. 

It seems that our current API `inserter:true|false` isn't enough. We have use cases that we would like to hide/show `source blocks` in the main inserter but being discoverable in the `slash inserter`, for example we want to hide `Query Loop` from the main inserter but show it's variations, and at the `slash inserter` to show the `Query Loop` block.

I don't have a clear way forward but I think we will need  involve `allowedBlocks` API (at least it's concept 😄 ) as well.

**Comment by getdave** [comment]
@adamziel this feels related to my efforts in https://github.com/WordPress/gutenberg/pull/37291

**Comment by adamziel** [comment]
Closing this PR, https://github.com/WordPress/gutenberg/pull/37157 did the trick.

**Pull request closed without merging**